Add missing numpy.random distributions

PierreBlancfat opened the pull request:
I took my inspiration from numpy/random/poisson.hpp. I made a flake8. 
Please do not hesitate to feedback about mistakes or omissions I could have done in this implementation.

serge-sans-paille commented on pythran/tests/test_numpy_random.py:
```diff
@@ -633,3 +633,69 @@ def numpy_random_bytes1(n):
+            from numpy.random import exponential
+            from numpy import var, mean
+            a = [exponential() for x in range(size)]
+            print(mean(a))
```
the print is probably useless here

serge-sans-paille commented on pythran/tests/test_numpy_random.py:
```diff
@@ -633,3 +633,69 @@ def numpy_random_bytes1(n):
+            scale = 2
+            a = exponential(scale, size)
+            print(mean(a))
+            print(var(a,ddof=0))
```
And here too,

serge-sans-paille commented:
+ you need to pass clang-format-3.6 on the source, and there seems to be output difference in the test. Either increase the number of iterations, or review your implementation wrt numpy one?

serge-sans-paille commented:
Cool! Let's wait until travis validates and merge that, then you can do a second blast of commits. If you can just update the commit history to reflect the change, either by squashing all changes in one commit and provide a decent commit message, or using one commit per distribution (more work on your side)

serge-sans-paille commented on pythran/tests/test_numpy_random.py:
```diff
@@ -633,3 +633,810 @@ def numpy_random_bytes1(n):
+            from numpy.random import standard_exponential
+            from numpy import var, mean
+            a = standard_exponential(size)
+            print(mean(a))
```
please remove these print in the test :-)

PierreBlancfat replied:
Yes sorry, I will remove them in the next commit, I forgot to do it :)

PierreBlancfat commented:
I don't understand why the five first travis jobs always fail, do you have an idea ?
Anyway I have squashed the previous commits into one, I push it here.

serge-sans-paille commented:
@PierreBlancfat : the error are available online, looks like you need an explicit cast, as in https://travis-ci.com/serge-sans-paille/pythran/jobs/229207532#L1793

serge-sans-paille commented:
@PierreBlancfat : still the same errors, the fix seems to be trivial:

```diff
-std::chi_squared_distribution<float> distribution{df};
+std::chi_squared_distribution<float> distribution{static_cast<float>(df)};
```

and similar changes where needed. Follow the link above!

serge-sans-paille commented on pythran/pythonic/include/numpy/random/chisquare.hpp:
```diff
@@ -0,0 +1,27 @@
+  namespace random
+  {
+    template <class pS>
+    types::ndarray<double, pS> chisquare(float df, pS const &shape);
```
s/float/double/ ?

serge-sans-paille commented on pythran/pythonic/numpy/random/gumbel.hpp:
```diff
@@ -0,0 +1,50 @@
+#include "pythonic/types/NoneType.hpp"
+#include "pythonic/types/tuple.hpp"
+#include "pythonic/utils/functor.hpp"
+#include <math.h>
```
you shouldn't need <math.h>. Use `xsimd::log`` instead

serge-sans-paille commented on pythran/pythonic/numpy/random/laplace.hpp:
```diff
@@ -0,0 +1,55 @@
+#include "pythonic/types/NoneType.hpp"
+#include "pythonic/types/tuple.hpp"
+#include "pythonic/utils/functor.hpp"
+#include <math.h>
```
<math.h>

serge-sans-paille commented on pythran/pythonic/numpy/random/logistic.hpp:
```diff
@@ -0,0 +1,51 @@
+#include "pythonic/types/NoneType.hpp"
+#include "pythonic/types/tuple.hpp"
+#include "pythonic/utils/functor.hpp"
+#include <math.h>
```
<math.h>

serge-sans-paille commented on pythran/pythonic/numpy/random/logseries.hpp:
```diff
@@ -0,0 +1,69 @@
+      double q, r, U, V;
+      double result;
+
+      r = log(1.0 - p);
```
would it make sense to use `log1p` instead?

serge-sans-paille commented on pythran/pythonic/numpy/random/logseries.hpp:
```diff
@@ -0,0 +1,69 @@
+          return 1;
+        }
+        U = std::uniform_real_distribution<double>{0., 1.}(details::generator);
+        q = 1.0 - exp(r * U);
```
and `expm1` here?

serge-sans-paille commented on pythran/pythonic/numpy/random/pareto.hpp:
```diff
@@ -0,0 +1,50 @@
+#include "pythonic/types/ndarray.hpp"
+#include "pythonic/types/tuple.hpp"
+#include "pythonic/utils/functor.hpp"
+#include <math.h>
```
math.h :-)

serge-sans-paille commented on pythran/pythonic/numpy/random/pareto.hpp:
```diff
@@ -0,0 +1,50 @@
+      types::ndarray<double, pS> result{shape, types::none_type()};
+      std::exponential_distribution<float> distribution{};
+      std::generate(result.fbegin(), result.fend(), [&]() {
+        return exp(distribution(details::generator) / a) - 1;
```
expm1 ?

serge-sans-paille commented on pythran/pythonic/numpy/random/power.hpp:
```diff
@@ -0,0 +1,50 @@
+      types::ndarray<double, pS> result{shape, types::none_type()};
+      std::exponential_distribution<double> distribution{};
+      std::generate(result.fbegin(), result.fend(), [&]() {
+        return pow(1 - exp(-distribution(details::generator)), 1. / a);
```
expm1

serge-sans-paille commented on pythran/pythonic/numpy/random/power.hpp:
```diff
@@ -0,0 +1,50 @@
+    double power(double a, types::none_type d)
+    {
+      return pow(
+          1 - exp(-std::exponential_distribution<double>{}(details::generator)),
```
same here

serge-sans-paille commented on pythran/pythonic/numpy/random/power.hpp:
```diff
@@ -0,0 +1,50 @@
+    {
+      types::ndarray<double, pS> result{shape, types::none_type()};
+      std::exponential_distribution<double> distribution{};
+      std::generate(result.fbegin(), result.fend(), [&]() {
```
you could just call power(a) here, right?

serge-sans-paille commented on pythran/pythonic/numpy/random/rayleigh.hpp:
```diff
@@ -0,0 +1,49 @@
+    {
+      types::ndarray<double, pS> result{array_shape, types::none_type()};
+      std::uniform_real_distribution<double> distribution{0., 1.};
+      std::generate(result.fbegin(), result.fend(), [&]() {
```
same here, instead of the lambda, use double rayleigh(double scale) defined below.

serge-sans-paille commented on pythran/pythonic/numpy/random/standard_gamma.hpp:
```diff
@@ -0,0 +1,42 @@
+
+    template <class pS>
+    types::ndarray<double, pS> standard_gamma(double s, pS const &shape)
+    {
```
There's a special way to define aliases in pythran, have a look to `include/numpy/diagonal.hpp` and `USING_FUNCTOR`

serge-sans-paille commented on pythran/pythonic/numpy/random/weibull.hpp:
```diff
@@ -0,0 +1,46 @@
+  {
+
+    template <class pS>
+    types::ndarray<double, pS> weibull(float a, pS const &shape)
```
float -> double ?

serge-sans-paille commented:
PlusI've add failure on travis for numpy_logistic0b (may be due to the random aspect of the test though)

PierreBlancfat commented:
I have fixed the casting and code formatting issues. Now cause tests are stochastic, one test about the logistic distribution have unfortunately failed. I changed the range of error of this test. I hope all tests will succeed !

PierreBlancfat commented:
> PlusI've add failure on travis for numpy_logistic0b (may be due to the random aspect of the test though)

I didn't see your message, I have just pushed again a version with a slightly change on the range of error of the test that failed.

serge-sans-paille commented:
Still failing for various test case :-/

paugier commented:
Bravo @PierreBlancfat !
